### ISSUE TITLE
Working very well on training but doing pretty poorly at inference

### ISSUE THREAD

**Issue by WaterKnight1998** [clovaai/donut#65]
Hi,

Thanks you for publishing this model :)

I want to use this model for Document Parsing. I have annotations for two kinds of pdf, 20 images per type.

At training it achieves very good edit distances 0.08, 0.0, 0.1, 0.13... But when I try it at inference the edit distance is very bad 0.8

In addition in those predictions I am seeing that in several predictions it outputs: `{"text_sequence": "word1 word1 word1 ... word1"}` where word1 is a random word repeated all over the place

Thanks in advance

**Comment by satheeshkatipomu** [comment]
what is the prompt you are using while inference? most probably it is messing up the json output. If you have fine-tuned base model for your task, try using `<s_REPLACE_YOUR_DATASET_NAME>` as prompt at inference time.

**Comment by WaterKnight1998** [comment]
Hi @satheeshkatipomu , thank you for answering.

I was following [this training tutorial](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/Donut/CORD/Quick_inference_with_DONUT_for_Document_Parsing.ipynb).

If you check his lightning module:
```python
def validation_step(self, batch, batch_idx, dataset_idx=0):
        pixel_values, labels, answers = batch
        batch_size = pixel_values.shape[0]
        # we feed the prompt to the model
        decoder_input_ids = torch.full((batch_size, 1), self.model.config.decoder_start_token_id, device=self.device)
        
        outputs = self.model.generate(pixel_values,
                                   decoder_input_ids=decoder_input_ids,
                                   max_length=max_length,
                                   early_stopping=True,
                                   pad_token_id=self.processor.tokenizer.pad_token_id,
                                   eos_token_id=self.processor.tokenizer.eos_token_id,
                                   use_cache=True,
                                   num_beams=1,
                                   bad_words_ids=[[self.processor.tokenizer.unk_token_id]],
                                   return_dict_in_generate=True,)
    
        predictions = []
        for seq in self.processor.tokenizer.batch_decode(outputs.sequences):
            seq = seq.replace(self.processor.tokenizer.eos_token, "").replace(self.processor.tokenizer.pad_token, "")
            seq = re.sub(r"<.*?>", "", seq, count=1).strip()  # remove first task start token
            predictions.append(seq)

        scores = list()
        for pred, answer in zip(predictions, answers):
            pred = re.sub(r"(?:(?<=>) | (?=</s_))", "", pred)
            # NOT NEEDED ANYMORE
            # answer = re.sub(r"<.*?>", "", answer, count=1)
            answer = answer.replace(self.processor.tokenizer.eos_token, "")
            scores.append(edit_distance(pred, answer) / max(len(pred), len(answer)))

            if self.config.get("verbose", False) and len(scores) == 1:
                print(f"Prediction: {pred}")
                print(f"    Answer: {answer}")
                print(f" Normed ED: {scores[0]}")

        return scores
```
He uses `decoder_input_ids = torch.full((batch_size, 1), self.model.config.decoder_start_token_id, device=self.device)` as decoder_input_ids instead of 

```python
    task_prompt = "<s_cord-v2>"
    decoder_input_ids = processor.tokenizer(task_prompt, add_special_tokens=False, return_tensors="pt").input_ids
```

If at inference time I use: `decoder_input_ids = torch.full((batch_size, 1), self.model.config.decoder_start_token_id, device=self.device)` the performance improves

**Comment by satheeshkatipomu** [comment]
did you fine-tune the model on your custom dataset? or trying to use the off the shelf `cord-v2` model?

**Comment by WaterKnight1998** [comment]
Yes, I finetuned the model with my custom dataset but I used that code @satheeshkatipomu

**Comment by satheeshkatipomu** [comment]
I see some changes compared to the code in this repo. but still I suggest checking your prompt at inference time if you think edit distance very less during training but very high while inference AND assuming your test images are similar to training data.

For example here in the fine-tuning notebook they hard corded start and end token as `<s_cord-v2>`. 
![prompt_token](https://user-images.githubusercontent.com/44122848/194034549-ae0100df-4bdf-4fd8-8a5a-56a159bda9e6.png)

Please check if it same during training and inference, If yes, sorry I am unable to help solve your problem.

**Comment by Whatever929** [comment]
I had the same issue as you, the model performed very well in validation during training, but very poorly during inference (the model just output gibberish).

Try to check ```added_tokens.json``` file in your trained model path. Your task prompt should be among the added tokens. For some reason, my task prompt was just ```<s_>``` instead of ```<s_{DATASET_NAME}>```. 
<img width="202" alt="added_json" src="https://user-images.githubusercontent.com/60956360/205490289-ecae2905-9b68-4fef-8491-f8a44d00242d.png">

**Comment by Mohtadrao** [comment]
I am having the same issue as it is giving very good accuracy and prediction during validation however it is giving garbish value while predicting the same image. My added tokens json is fine. Kindly help. @WaterKnight1998 @satheeshkatipomu @outday29

**Comment by CarlosSerrano88** [comment]
@Mohtadrao @WaterKnight1998 have you founded a solución? I have the same error. Thanks

**Comment by Mohtadrao** [comment]
@CarlosSerrano88 yeah. My accuracy was increase when my number of epochs along with training data was increased. Also need to adjust the config file accordingly.

**Comment by Ruxin124** [comment]
during training at the begining epoches the predictions are like this, is it normal? i just need to wait 200 epoches finished ot is there some thing wrong? 
![image](https://github.com/user-attachments/assets/6642a72b-cc07-4241-a44e-7084dd13cd90)

**Comment by Ruxin124** [comment]
> @Mohtadrao @WaterKnight1998 have you founded a solución? I have the same error. Thanks
Same problem here, predictions are all (s) (s) (s) .... 
Have someone solved the problem ? 
![image](https://github.com/user-attachments/assets/f2691e99-5094-4296-acdd-fb3170782974)

**Comment by northwindapps** [comment]
tried prediction with data on a custom data huggingface repo and could confirmed that the dataset name in the "added_tokens.json" which get produced after a training.


<img width="351" alt="image" src="https://github.com/user-attachments/assets/ed275a97-5aab-445e-ad70-388c07b0cef6">

set the datasetname after "<s_ ....>"
`output` = model.inference(image=image, prompt="<s_donut-jp-bill>")`

**Comment by Mohtadrao** [comment]
@Ruxin124 This is normal error. You just need to train, I dont know how much of data you have but considering my data that was aprox 7-8k images. For 45 epochs it was already giving me 90% accuracy. Also need to write config file accurately according to your data.
